File: bin/base/scripts/screens/test_scroll_screen.lua
class "TestScrollScreen" (Screen)

function TestScrollScreen:init()
    self.field_width = 80
    self.field_height = 50
    self.tile_width = 128
    self.tile_height = 64
    self.tile_sprite = "test_tile_spr"

    local screen_width = Engine.getScreenWidth()
    local screen_height = Engine.getScreenHeight()

    self.scroll_cnt = ScrollContainer("scrollCnt")
    self.scroll_cnt:setRect(0, 0, screen_width, screen_height)
    self.scroll_cnt:setScrollSpeed(1000)
    self:attach(self.scroll_cnt)

    local left_area = Area("scrollLeftArea")
    left_area:setRect(0, 0, 20, 700)
    left_area:setAlignment("LEFT|MIDDLE", 0, 0)
    left_area:addCallback("MouseOver", self.scrollTo, {self, "Left", true})
    left_area:addCallback("MouseLeft", self.scrollTo, {self, "Left", false})
    self:attach(left_area)

    local up_area = Area("scrollUpArea")
    up_area:setRect(0, 0, 1000, 20)
    up_area:setAlignment("CENTER|TOP", 0, 0)
    up_area:addCallback("MouseOver", self.scrollTo, {self, "Up", true})
    up_area:addCallback("MouseLeft", self.scrollTo, {self, "Up", false})
    self:attach(up_area)

    local down_area = Area("scrollDownArea")
    down_area:setRect(0, 0, 1000, 20)
    down_area:setAlignment("CENTER|BOTTOM", 0, 0)
    down_area:addCallback("MouseOver", self.scrollTo, {self, "Down", true})
    down_area:addCallback("MouseLeft", self.scrollTo, {self, "Down", false})
    self:attach(down_area)

    local right_area = Area("scrollRightArea")
    right_area:setRect(0, 0, 20, 700)
    right_area:setAlignment("RIGHT|MIDDLE", 0, 0)
    right_area:addCallback("MouseOver", self.scrollTo, {self, "Right", true})
    right_area:addCallback("MouseLeft", self.scrollTo, {self, "Right", false})
    self:attach(right_area)

    local btn = Button("backBtn")
    btn:setText("Exit")
    btn:setFont("system_15_fnt")
    btn:setRect(0, 0, 64, 64)
    btn:setAlignment("RIGHT|BOTTOM", 0, 0)
    btn:setTextAlignment("CENTER|MIDDLE")
    btn:setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr")
    btn:addCallback("MouseUp_Left", self.onBackBtnClick, self)
    btn:setColour("red")
    self:attach(btn)

    btn = Button("jumpBtn")
    btn:setText("Jump")
    btn:setFont("system_15_fnt")
    btn:setRect(0, 0, 64, 64)
    btn:setAlignment("RIGHT|BOTTOM", 0, -64)
    btn:setTextAlignment("CENTER|MIDDLE")
    btn:setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr")
    btn:setColour("red")
    btn:addCallback("MouseUp_Left", self.onJumpBtnClick, self)
    self:attach(btn)

    btn = Button("scrollBtn")
    btn:setText("Scroll")
    btn:setFont("system_15_fnt")
    btn:setRect(0, 0, 64, 64)
    btn:setAlignment("RIGHT|BOTTOM", 0, -128)
    btn:setTextAlignment("CENTER|MIDDLE")
    btn:setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr")
    btn:setColour("red")
    btn:addCallback("MouseUp_Left", self.onScrollBtnClick, self)
    self:attach(btn)

    self:addScrollContent()
end

function TestScrollScreen.scrollTo(params)
    local screen = params[1]
    local direction = params[2]
    local value = params[3]
    if (screen and screen.scroll_cnt) then
        screen.scroll_cnt:enableScroll(direction, value)
    end
end

function TestScrollScreen:onBackBtnClick()
    Screens.load("LoadingScreen", "TestScreen")
end

function TestScrollScreen:onJumpBtnClick()
    if (self.scroll_cnt) then
        self.scroll_cnt:jumpTo(1024, 768)
    end
end

function TestScrollScreen:onScrollBtnClick()
    if (self.scroll_cnt) then
        if (not self.scroll_cnt:isScrolling()) then
            self.scroll_cnt:scrollTo(0, 0)
        end
    end
end

function TestScrollScreen:addScrollContent()
    if (self.scroll_cnt) then
        local content_width = 0
        local content_height = 0
        for i = 0, self.field_width - 1 do
            for j = 0, self.field_height - 1 do
                local is_odd = 1 == j % 2

                local offset_x = is_odd and (self.tile_width / 2) or 0
                local x = self.tile_width * (self.field_width - 1 - i) + offset_x
                local y = self.tile_height / 2 * j

                local img = Image()
                img:setRect(x, y, self.tile_width, self.tile_height)
                img:setSprite(self.tile_sprite)
                self.scroll_cnt:attach(img)

                if (x + self.tile_width > content_width) then
                    content_width = x + self.tile_width
                end
                if (y + self.tile_height) then
                    content_height = y + self.tile_height
                end
            end
        end
        self.scroll_cnt:setContentRect(0, 0, content_width, content_height)
    end
end

File: src/test_scroll_screen.cpp
#include "test_scroll_screen.h"

#include "system_tools.h"

namespace redrevolt
{
TestScrollScreen::TestScrollScreen(const std::string & id)
    : Screen(id)
    , m_fieldWidth(80)
    , m_fieldHeight(50)
    , m_tileWidth(128)
    , m_tileHeight(64)
    , m_tileSpr("test_tile_spr")
    , m_scrollCnt(nullptr)
{
    SystemTools * systemTools = new SystemTools("systemTools");
    attach(systemTools);

    const int screenWidth = stren::Engine::getScreenWidth();
    const int screenHeight = stren::Engine::getScreenHeight();

    m_scrollCnt = new stren::ScrollContainer("scrollCnt");
    m_scrollCnt->setRect(0, 0, screenWidth, screenHeight);
    m_scrollCnt->setScrollSpeed(1000);
    attach(m_scrollCnt);

    stren::Area * leftArea = new stren::Area("scrollLeftArea");
    leftArea->setRect(0, 0, 20, 700);
    leftArea->setAlignment("LEFT|MIDDLE", 0, 0);
    leftArea->addCallback("MouseOver", this, &TestScrollScreen::scrollLeftOn);
    leftArea->addCallback("MouseLeft", this, &TestScrollScreen::scrollLeftOff);
    attach(leftArea);
    
    stren::Area * upArea = new stren::Area("scrollUpArea");
    upArea->setRect(0, 0, 1000, 20);
    upArea->setAlignment("CENTER|TOP", 0, 0);
    upArea->addCallback("MouseOver", this, &TestScrollScreen::scrollUpOn);
    upArea->addCallback("MouseLeft", this, &TestScrollScreen::scrollUpOff);
    attach(upArea);
    
    stren::Area * downArea = new stren::Area("scrollDownArea");
    downArea->setRect(0, 0, 1000, 20);
    downArea->setAlignment("CENTER|BOTTOM", 0, 0);
    downArea->addCallback("MouseOver", this, &TestScrollScreen::scrollDownOn);
    downArea->addCallback("MouseLeft", this, &TestScrollScreen::scrollDownOff);
    attach(downArea);
    
    stren::Area * rightArea = new stren::Area("scrollRightArea");
    rightArea->setRect(0, 0, 20, 700);
    rightArea->setAlignment("RIGHT|MIDDLE", 0, 0);
    rightArea->addCallback("MouseOver", this, &TestScrollScreen::scrollRightOn);
    rightArea->addCallback("MouseLeft", this, &TestScrollScreen::scrollRightOff);
    attach(rightArea);

    stren::Button * btn = new stren::Button("backBtn");
    btn->setText("Exit");
    btn->setFont("system_15_fnt");
    btn->setRect(0, 0, 64, 64);
    btn->setAlignment("RIGHT|BOTTOM", 0, 0);
    btn->setTextAlignment("CENTER|MIDDLE");
    btn->setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr");
    btn->addCallback("MouseUp_Left", this, &TestScrollScreen::onBackBtnClick);
    btn->setColour("red");
    attach(btn);

    btn = new stren::Button("jumpBtn");
    btn->setText("Jump");
    btn->setFont("system_15_fnt");
    btn->setRect(0, 0, 64, 64);
    btn->setAlignment("RIGHT|BOTTOM", 0, -64);
    btn->setTextAlignment("CENTER|MIDDLE");
    btn->setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr");
    btn->addCallback("MouseUp_Left", this, &TestScrollScreen::onJumpBtnClick);
    btn->setColour("red");
    attach(btn);

    btn = new stren::Button("scrollBtn");
    btn->setText("Scroll");
    btn->setFont("system_15_fnt");
    btn->setRect(0, 0, 64, 64);
    btn->setAlignment("RIGHT|BOTTOM", 0, -128);
    btn->setTextAlignment("CENTER|MIDDLE");
    btn->setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr");
    btn->addCallback("MouseUp_Left", this, &TestScrollScreen::onScrollBtnClick);
    btn->setColour("red");
    attach(btn);

    addScrollContent();
}

TestScrollScreen::~TestScrollScreen()
{
}

void TestScrollScreen::scrollTo(const int direction, const bool value)
{
    if (m_scrollCnt)
    {
        m_scrollCnt->scrollTo(direction, value);
    }
}

void TestScrollScreen::scrollUpOn()
{
    scrollTo(stren::ScrollContainer::Up, true);
}

void TestScrollScreen::scrollUpOff()
{
    scrollTo(stren::ScrollContainer::Up, false);
}

void TestScrollScreen::scrollLeftOn()
{
    scrollTo(stren::ScrollContainer::Left, true);
}

void TestScrollScreen::scrollLeftOff()
{
    scrollTo(stren::ScrollContainer::Left, false);
}

void TestScrollScreen::scrollRightOn()
{
    scrollTo(stren::ScrollContainer::Right, true);
}

void TestScrollScreen::scrollRightOff()
{
    scrollTo(stren::ScrollContainer::Right, false);
}

void TestScrollScreen::scrollDownOn()
{
    scrollTo(stren::ScrollContainer::Down, true);
}

void TestScrollScreen::scrollDownOff()
{
    scrollTo(stren::ScrollContainer::Down, false);
}

void TestScrollScreen::onBackBtnClick(Widget * sender)
{
    stren::SwitchScreenCommand command;
    command.setScreen("LoadingScreen");
    command.setNextScreen("TestScreen");
    command.execute();
 }

void TestScrollScreen::onJumpBtnClick(Widget * sender)
{
    if (m_scrollCnt)
    {
        m_scrollCnt->jumpTo(1024, 768);
    }
}

void TestScrollScreen::onScrollBtnClick(Widget * sender)
{
    if (m_scrollCnt)
    {
        if (!m_scrollCnt->isScrolling())
        {
            m_scrollCnt->scrollTo(0, 0);
        }
    }
}

void TestScrollScreen::addScrollContent()
{
    if (m_scrollCnt)
    {
        int contentWidth(0);
        int contentHeight(0);
        for (int i = 0; i < m_fieldWidth; ++i)
        {
            for (int j = 0; j < m_fieldHeight; ++j)
            {
                const bool isOdd = 1 == (j & 1);
                const int offsetX = isOdd ? (m_tileWidth * 0.5) : 0;
                const int x = m_tileWidth * (m_fieldWidth - 1 - i) + offsetX;
                const int y = m_tileHeight * 0.5 * j;

                stren::Image * img = new stren::Image();
                img->setRect(x, y, m_tileWidth, m_tileHeight);
                img->setSprite(m_tileSpr);
                m_scrollCnt->attach(img);

                if (x + m_tileWidth > contentWidth)
                {
                    contentWidth = x + m_tileWidth;
                }
                if (y + m_tileHeight > contentHeight)
                {
                    contentHeight = y + m_tileHeight;
                }
            }
        }
        m_scrollCnt->setContentRect(0, 0, contentWidth, contentHeight);
    }
}
} // redrevolt

